core/src/impl/Kokkos_Profiling.cpp

```diff
@@ -97,13 +108,26 @@ bool eventSetsEqual(const EventSet& l, const EventSet& r) {
          l.end_fence == r.end_fence && l.sync_dual_view == r.sync_dual_view &&
          l.modify_dual_view == r.modify_dual_view &&
          l.declare_metadata == r.declare_metadata &&
+         l.request_responses == r.request_responses &&
+         l.transmit_actions == r.transmit_actions &&
          l.declare_input_type == r.declare_input_type &&
          l.declare_output_type == r.declare_output_type &&
          l.end_tuning_context == r.end_tuning_context &&
          l.begin_tuning_context == r.begin_tuning_context &&
          l.request_output_values == r.request_output_values &&
          l.declare_optimization_goal == r.declare_optimization_goal;
 }
+template <typename Callback, typename... Args>
+void call_kokkos_callback(const Callback callback, bool ever_needs_fence,
+                          Args... args) {
+  if (callback != nullptr) {
+    if ((ever_needs_fence) && (Kokkos::Tools::Experimental::tool_requirements
+                                   .requires_global_fencing)) {
+      Kokkos::fence();
+    }
+    (*callback)(args...);
+  }
+}
```

Comment:
I think this would help optimize the call:

```suggestion
#if __cplusplus >= 201703L
#        define KOKKOS_IF_CONSTEXPR if constexpr
#else
#        define KOKKOS_IF_CONSTEXPR if
#endif
template <typename Callback, typename EverNeedsFence, typename... Args>
void call_kokkos_callback(const Callback callback, EverNeedsFence,
                          Args... args) {
  if (callback != nullptr) {
    // a macro for if constexpr is needed
    KOKKOS_IF_CONSTEXPR (EverNeedsFence::value) {
      if (Kokkos::Tools::Experimental::tool_requirements.requires_global_fencing) {
        Kokkos::fence();
      }
    }
    (*callback)(args...);
  }
}
```

Usage:

```cpp
Kokkos::Tools::Experimental::call_kokkos_callback(
      Kokkos::Tools::Experimental::current_callbacks.begin_parallel_for, std::true_type{},
      kernelPrefix.c_str(), devID, kernelID);
```

Reply:
Even without this change, it would _likely_ result in the same code after optimization but this would give a much stronger hint in C++14 and ensure it in C++17

Reply:
@dhollman @dalg24 : do we have a preferred way to do `KOKKOS_IF_CONSTEXPR` already? Runtime `if` before C++17, constexpr after?

Reply:
I'm going to mark this one resolved, based on the compromise we made earlier. I've changed to a true_type/false_type, but no constexpr